src/main/java/picocli/CommandLine.java

```diff
@@ -11726,33 +11719,34 @@ void validate(CommandLine commandLine) {
 
             private void failGroupMultiplicityExceeded(List<ParseResult.GroupMatch> groupMatches, CommandLine commandLine) {
                 Map<ArgGroupSpec, List<List<ParseResult.GroupMatch>>> matchesPerGroup = new LinkedHashMap<ArgGroupSpec, List<List<GroupMatch>>>();
-                String msg = "";
+                StringBuilder msgBuilder = new StringBuilder();
                 for (ParseResult.GroupMatch match : groupMatches) {
-                    if (msg.length() > 0) { msg += " and "; }
-                    msg += match.toString();
+                    if (msgBuilder.length() > 0) { msgBuilder.append(" and "); }
+                    msgBuilder.append(match);
                     Map<ArgGroupSpec, GroupMatchContainer> subgroups = match.matchedSubgroups();
-                    for (ArgGroupSpec group : subgroups.keySet()) {
+                    for (Map.Entry<ArgGroupSpec, GroupMatchContainer> groupEntry : subgroups.entrySet()) {
+                        ArgGroupSpec group = groupEntry.getKey();
                         if (group.validate()) { // don't raise errors for non-validating groups: https://github.com/remkop/picocli/issues/810
-                            addValueToListInMap(matchesPerGroup, group, subgroups.get(group).matches());
+                            addValueToListInMap(matchesPerGroup, group, groupEntry.getValue().matches());
```

Comment:
I like the previous iteration style better, because it gives shorter types (`ArgGroupSpec` vs `Map.Entry<ArgGroupSpec, GroupMatchContainer>`) and more meaningful variable names.